benchmarks/report_bayesmark.py

```diff
@@ -1,55 +1,255 @@
+        wins: Dict[str, int] = defaultdict(int)
+        alpha = DewanckerRanker.pick_alpha(report)
+        for metric in self._metrics:
+            samples = report.sample_performance(metric)
+            for optim_a, optim_b in itertools.permutations(samples, 2):
+                _, p_val = mannwhitneyu(samples[optim_a], samples[optim_b], alternative="less")
+                if p_val < alpha:
+                    wins[optim_a] += 1
+
+            all_wins = [wins[optimizer] for optimizer in report.optimizers]
+            no_ties = len(all_wins) == len(np.unique(all_wins))
+            if no_ties:
+                break
+
+        wins = {optimzier: wins[optimzier] for optimzier in report.optimizers}
+        self._set_ranking(wins)
+        self._set_borda(wins)
 
 
 class BayesmarkReportBuilder:
     def __init__(self) -> None:
 
-        self._solvers: Counter = Counter()
-        self._datasets: Counter = Counter()
-        self._models: Counter = Counter()
+        self._solvers: Set[str] = set()
+        self._datasets: Set[str] = set()
+        self._models: Set[str] = set()
+        self._firsts: Dict[str, int] = defaultdict(int)
+        self._borda: Dict[str, int] = defaultdict(int)
+        self._metric_precedence = str()
         self._problems_counter = 1
         self._problems_body = io.StringIO()
 
-    def add_problem(self, name: str, scores: Dict[str, float]) -> "BayesmarkReportBuilder":
+    def set_precedence(self, metrics: List[BaseMetric]) -> None:
+
+        self._metric_precedence = " -> ".join([m.name for m in metrics])
+
+    def add_problem(
+        self,
+        name: str,
+        report: PartialReport,
+        ranking: DewanckerRanker,
+        metrics: List[BaseMetric],
+    ) -> "BayesmarkReportBuilder":
 
         if self._problems_body.closed:
             self._problems_body = io.StringIO()
 
-        problem_title = f"### ({self._problems_counter}) Problem: {name}\n"
-        self._problems_body.write(problem_title)
-        self._problems_body.write(_TABLE_HEADER)
+        problem_header = f"### ({self._problems_counter}) Problem: {name}" + _LINE_BREAK
+        self._problems_body.write(problem_header)
+        metric_names = [f"{m.name} (avg +- std) |" for m in metrics]
+        self._problems_body.write(
+            "".join([_LINE_BREAK, _TABLE_HEADER, *metric_names, _LINE_BREAK])
+        )
+        metric_format = ["---:|" for _ in range(len(metrics))]
+        self._problems_body.write("".join([_HEADER_FORMAT, *metric_format, _LINE_BREAK]))
+
+        positions = np.abs(ranking.borda - (max(ranking.borda) + 1))
+        for pos, solver in zip(positions, ranking.solvers):
+            self._solvers.add(solver)
+            row_buffer = io.StringIO()
+            row_buffer.write(f"|{pos}|{solver}|")
+            for metric in metrics:
+                mean, variance = report.summarize_solver(solver, metric)
+                row_buffer.write(f"{mean:.5f} +- {np.sqrt(variance):.5f}|")
```

Comment:
Oh, btw, currently each metric is rendered with fixed number of decimal places,which is not true to `kurobako`, where this changes depending on metric. I initially planned to implement that by adding another class attribute to each metric (besides name) which would make it very convenient to reference here but, well... I forgot :facepalm:. Can do it now, or save it for some neat follow-up, depends how do you feel about that @HideakiImamura.

Reply:
To simplify the change, I think it is good to say that this should be a follow-up.